### ISSUE TITLE
[JavaSpring] API-levle consumes should only affect relevant verbs

### ISSUE THREAD

**Issue by slarti-b** [swagger-api/swagger-codegen#7575]
##### Description

If I include the produces/consumes declarations at the top of the yaml spec then they are global and apply to every endpoint in the API.  This is why I want to include them there - it means I can define them once and that's it.

However, that means it also applies the `consumes` to GET and DELETE requests, which don't take a body.  I just wrote my first API using swagger, and my colleague who is consuming it in the first implementation can't call the GET endpoints.  He's consuming it from a .NET application and the .NET httpClient throws an exception if you try to set a Content-Type on a GET request!  But without it, my API says not found because it doesn't have a compatible Content-Type header.  It might be possible to work around with an ugly reflection hack, but we don't don't want to go there!

I know I can simply specify the consumes at the endpoint level where relevant, but then I repeat it.  

##### Swagger-codegen version

2.3.0

##### Swagger declaration file content or url

Example spec
```yaml
swagger: '2.0'
host: api.example.com
basePath: /
schemes:
  - https
consumes:
  - application/json
  - application/xml
produces:
  - application/json
  - application/xml
tags:
  - name: foo
paths:
  /foo:
    get:
      operationId: getFoo
      tags:
        - foo      
      responses:
        '200':
          description: OK
    post:
      operationId: postFoo
      tags:
        - foo      
      responses:
        '200':
          description: OK
    put:
      operationId: putFoo
      tags:
        - foo      
      responses:
        '200':
          description: OK
    delete:
      operationId: deleteFoo
      tags:
        - foo      
      responses:
        '200':
          description: OK       
```

Produces this Java code
```java
@Api(value = "foo", description = "the foo API")
public interface FooApi {

    FooApiDelegate getDelegate();

    @ApiOperation(value = "", nickname = "deleteFoo", notes = "", tags={ "foo", })
    @ApiResponses(value = { 
        @ApiResponse(code = 200, message = "OK") })
    @RequestMapping(value = "/foo",
        produces = { "application/json", "application/xml" }, 
        consumes = { "application/json", "application/xml" },
        method = RequestMethod.DELETE)
    default CompletableFuture<ResponseEntity<Void>> deleteFoo() {
        return getDelegate().deleteFoo();
    }


    @ApiOperation(value = "", nickname = "getFoo", notes = "", tags={ "foo", })
    @ApiResponses(value = { 
        @ApiResponse(code = 200, message = "OK") })
    @RequestMapping(value = "/foo",
        produces = { "application/json", "application/xml" }, 
        consumes = { "application/json", "application/xml" },
        method = RequestMethod.GET)
    default CompletableFuture<ResponseEntity<Void>> getFoo() {
        return getDelegate().getFoo();
    }


    @ApiOperation(value = "", nickname = "postFoo", notes = "", tags={ "foo", })
    @ApiResponses(value = { 
        @ApiResponse(code = 200, message = "OK") })
    @RequestMapping(value = "/foo",
        produces = { "application/json", "application/xml" }, 
        consumes = { "application/json", "application/xml" },
        method = RequestMethod.POST)
    default CompletableFuture<ResponseEntity<Void>> postFoo() {
        return getDelegate().postFoo();
    }


    @ApiOperation(value = "", nickname = "putFoo", notes = "", tags={ "foo", })
    @ApiResponses(value = { 
        @ApiResponse(code = 200, message = "OK") })
    @RequestMapping(value = "/foo",
        produces = { "application/json", "application/xml" }, 
        consumes = { "application/json", "application/xml" },
        method = RequestMethod.PUT)
    default CompletableFuture<ResponseEntity<Void>> putFoo() {
        return getDelegate().putFoo();
    }

}
```
##### Command line used for generation

Using swagger-codegen-maven-plugin:2.3.0 and swagger codegen 2.3.0
Using options:
- serializableModel: true
- dateLibrary: java8
- java8: true
- library: spring-boot
- async: true
- delegatePattern: true
- useOptional: true

##### Steps to reproduce

Generate the above spec for spring-boot

##### Suggest a fix/enhancement

I think there should be an option (at least, if not make it the default behaviour) to only generate a `consumes`  in the `@RequestMapping` annotation for POST, PUT and PATCH and not for DELETE, GET, etc.

I am aware there  is some discussion around if body is allowed in DELETE for example.  The spec doesn't forbid it, but says it is undefined behaviour.  My own experience is that, in addition to the GET issue in .NET) our API Gateway (WS02) strips out the body from DELETE, so I would argue they should be avoid.  However, I guess it needs to be an option - should it be default to generate consumes for all (as today) or default to only do it for those allowed?  In terms of defining what's !allowed" I propose using the MDN definitions (https://developer.mozilla.org/en-US/docs/Web/HTTP/Methods) - each method includes a table specifying "Request has body"  as Yes/No.

I'd even be happy to try and do this, but wanted some feedback if it's generally considered a good idea or not first!

**Comment by cbornet** [comment]
I guess we could prevent setting the consumes if there’s no request body. Same for response body and produces.

**Comment by cbornet** [comment]
I wouldn’t do it based on http verbs though

**Comment by slarti-b** [comment]
Yeah, that's maybe a good option.  Achieves the same result I guess.  I was thinking HTTP verbs since that seems to be where the client restrictions are coming from so far, but the main objective is to say "my api understands these formats" at a global/header level and then have it skip it where it causes a problem by not being relevant.  Would still need to be an option I assume.

**Comment by slarti-b** [comment]
@cbornet : I've been thinking about your comment a bit.  I do see the point that it reflects the definition more automatically.  My only concern would be that it makes the _routing_ dependent on the _current_ definition.

For example, suppose I have an API definition which produces JSON or XML (in the head) and includes a path `GET /foo/{id}` which initially only returns a status code to say if it exists or not.  It doesn't have a body, so per your suggestion it would not  define a `produces` for that path. 

The client would then be free to call it without any `Accepts` headers - or with `Accepts: text/plain` or whatever they liked.  It wouldn't matter, it would still find that endpoint.  

Now suppose I expand this to return a body.  It's still backwardly compatible, in that the status codes returned still have the same meanings but it's returning extra (optional) info.  Problem is that the client who was calling it without any `Accepts` headers (or with `Accepts: text/plain`, etc.) will now get _404 Not Found_, because the route now only matches if you have `Accepts` as either `application/json` or `application/xml`.

I guess my initial thought to use the HTTP Methods was motivated by two factors:

1. It's based on the http spec not the actual API, so if you break it, you're probably breaking other stuff anyway
2. It forces the APi author (who uses the setting - those who don't get what they get today) to follow a stricter/more restrictive interpretation of the http spec, and therefore to be more compatible with a wider range of clients and gateways.  Especially for people (like me) fairly new to writing REST APIs and not experienced yet with the API manager, etc. it might help steer them to avoid some of the pitfalls I've fallen into.

Note that I'm talking here about using the codegen to generate server-stubs, not clients.  The clients should probably add the headers if they can, since they can't hurt there.

That said, I acknowledge that my example might be a bit contrived/unlikely/theoretical and wouldn't be totally opposed to doing it your way, but would appreciate some comments on this scenario - and on why you think following the HTTP verbs is a bad idea?

**Comment by cbornet** [comment]
> Now suppose I expand this to return a body. It's still backwardly compatible, in that the status codes returned still have the same meanings but it's returning extra (optional) info. Problem is that the client who was calling it without any Accepts headers (or with Accepts: text/plain, etc.) will now get 404 Not Found, because the route now only matches if you have Accepts as either application/json or application/xml.

If you expand your endpoint to return a body, you need to specify its content-type so if your client doesn't accept it, yes that's a breaking change but you can't do anything about it. Note that if your client doesn't care about the returned content-type, it should set "Accepts: */*". That's what is done by curl for instance.

**Comment by cbornet** [comment]
> why you think following the HTTP verbs is a bad idea?

Because swagger doesn't put any constraints on the HTTP verbs. So it's cleaner to check if there's a body IMO.

**Comment by slarti-b** [comment]
OK.  And I guess this would need to be controlled by a flag too?  Otherwise it'd be breaking.  Will look at it when I get time.

**Comment by cbornet** [comment]
If it’s a fix, I don’t consider it as a breaking change so the PR can be done against master.
